### ISSUE TITLE
Learning BMSAIT for x27 and leds.

### ISSUE THREAD

**Issue by ECVLance** [HummerX/BMSAIT#9]
Hi,
I'm learning to use BMSAIT trying to find a software that woul allow me to control lights and x27 motors on BMS and DCS without the need to reprogram de Arduino board.
I'm reading the manual, and aparenty I had a succesful test using the Arduino sketch for the StepperX27 (the motor spins correctly). But i want to try pluggin two x27 stepper motors. From what i understand, i must add the DATA VARIABLES to the "BMSAIT_UserConfig.h" file, like this:

```
  _Datenfeld datenfeld[]=
    {
      //Description ID    DT    OT    target Ref2 Ref3 Ref4 Ref5  RQ   IV
        {"RPM",   "0341",  'i',   51,     2,    0,   0,   0,   0,   "", "0"}     //Variable - RPM
        ,{"FTIT",   "0261",  'i',   51,     3,    0,   0,   0,   0,   "", "0"}     //Variable - FTIT
    }; 
```
  const byte VARIABLENANZAHL = sizeof(datenfeld)/sizeof(datenfeld[0]);_ 
  
  And also add the pins to where the x27 are conected to the "BMSAIT_StepperX27.h" file, like this:

```
StepperdataX27 stepperdataX27[] =
{
  //  {PIN1 PIN2 PIN3 PIN4}    arc         last
    { {  2,   3,   4,   5   }, 315*3 ,      0   }  // RPM
      {  6,   7,   8,   9   }, 315*3 ,      0   }  // FTIT
};
```

But what i don't understand, and can not find, is how the BMSAIT app know which data of the shared memory is used to control each motor. I can't find where is the "link" that makes the RPM data is used to control the motor conected to pins 2 up to 5. The same with the FTIT data.

I know a little about programing Arduino, but i'm still a beginner.

**Comment by HummerX** [comment]
Hi ECVLance, 
you are right on track!
You understood that it is required to define data both in the UserConfig and the motor module.
In the motor module, you put in hardware related data (the Pin connections between the motor and the Arduino). 
The software related part (what data to use) is set up in the UserConfig. There, you put in a line for each data variable to use. The data to be used is defined by the four-digit ID in the second column of that data table. 

The entries of the tables are linked by the  "target" value in the UserConfig data. I.e. in your example above, the RPM value will be linked to the third line  in the motor module data block (first line is 0, second line is 1, third line is 3...). As you only have to lines there, the "target" value should be either 0 or 1.
I see that this is missing in the documentation. I will add this in the next update.

The link to the shared mem is programmed in the BMSAIT Windows app. For each data variable, I coded the link to a specific data field in the BMS SharedMem and the DCS data container.

I am far from done with the DCS integration. If you would specify the aircraft you want to use and which LED/motors you are trying to link, I can check if I already coded the link to DCS for these data fields.

**Comment by ECVLance** [comment]
Also, on another related/unralated subject.
The TWP, with the exception of the "LAUNCH" button that haves two lights that work together and flash with a missile launch, or on a test. All the other buttons have two lights that work independently. Most of them have one that is on with power present, and the other is on when the buttons is pressed or some conditiont is met.
But seeing the "BMSAIT-Variablen.csv" i only see six lights on the CMDS/RWR group for the TWP (ID 1541 up to 1546). They are the power light of each button? Or the other light?.

**Comment by ECVLance** [comment]
> ...
> The entries of the tables are linked by the "target" value in the UserConfig data. I.e. in your example above, the RPM value will be linked to the third line in the motor module data block (first line is 0, second line is 1, third line is 3...). As you only have to lines there, the "target" value should be either 0 or 1. I see that this is missing in the documentation. I will add this in the next update.
> ....

Well, thaks a lot for the help. I corrected the code as you said.

"BMSAIT_UserConfig.h" file is like this:
```
  Datenfeld datenfeld[]=
    {
      //Description ID    DT    OT    target Ref2 Ref3 Ref4 Ref5  RQ   IV
       //{"ENGFI",  "1506", 'b',  12,    0,     2,   4,   0,   0,   "",  "F"}     //Example Variable 0 - Right Eyebrow Engine Fire
      {"RPM",   "0341",  'i',   51,     0,    0,   0,   0,   0,   "", "0"}     //Variable - RPM
      ,{"FTIT",   "0261",  'i',   51,     1,    0,   0,   0,   0,   "", "0"}     //Variable - FTIT
    }; 
  const byte VARIABLENANZAHL = sizeof(datenfeld)/sizeof(datenfeld[0]); 
```

And "BMSAIT_StepperX27.h" file is like this:
```
StepperdataX27 stepperdataX27[] =
{
  //  {PIN1 PIN2 PIN3 PIN4}    arc         last
    { {  23,  25,  27,   29   }, 315*3 ,      0   }  // RPM
    { {  47,  49,  51,   53   }, 315*3 ,      0   }  // example: FTIT
};
```

This gave me a compile error.
_E:\......\BMSAIT_Vanilla\BMSAIT_StepperX27.h:26:5: error: expected '}' before '{' token
    { {  47,  49,  51,   53   }, 315*3 ,      0   }  // example: FTIT
     ^

E:\.....\BMSAIT_Vanilla\BMSAIT_StepperX27.h:26:5: error: expected ',' or ';' before '{' token
E:\....\BMSAIT_Vanilla\BMSAIT_StepperX27.h:27:1: error: expected declaration before '}' token
 };
 ^
exit status 1
Compilation error: expected '}' before '{' token_

The i changed the motor module code of the "BMSAIT_StepperX27.h" file like this:
```
StepperdataX27 stepperdataX27[] =
{
  //  {PIN1 PIN2 PIN3 PIN4}    arc         last
    { {  23,  25,  27,   29   }, 315*3 ,      0   }  // RPM
   **,** { {  47,  49,  51,   53   }, 315*3 ,      0   }  // example: FTIT
};
```
Added the "comma" in the begining of the FTIT line, and i got this compile error message:
_E:\....\BMSAIT_Vanilla\BMSAIT_StepperX27.h:34:1: error: could not convert '<brace-enclosed initializer list>()' from '<brace-enclosed initializer list>' to 'SwitecX25'
 };
 ^
exit status 1
Compilation error: could not convert '<brace-enclosed initializer list>()' from '<brace-enclosed initializer list>' to 'SwitecX25'_

And now i'm stuck. Don't know what is wrong. With a single line it compiles fine, but with two lines i got compile error.

**Comment by ECVLance** [comment]
Aparently the problem was on the SwitecX25 array. I had to add another line on the Switc25 array,

```
StepperdataX27 stepperdataX27[] =
{
  //  {PIN1 PIN2 PIN3 PIN4}    arc         last
    { {  23,  25,  27,   29   }, 315*3 ,      0   }  // RPM
    ,{ {  47,  49,  51,   53   }, 315*3 ,      0   }  // example: FTIT
};

const int stepperzahlX27 = sizeof(stepperdataX27)/sizeof(stepperdataX27[0]);

SwitecX25 stepperX27[stepperzahlX27]=
{
  SwitecX25(945,0,0,0,0),
  SwitecX25(945,0,0,0,0)
}; 
```
Like this, i dont have a compilation error. Although i'm not sure if the two lines of the "SwitecX25" array are correctly coded.

**Comment by HummerX** [comment]
Yes, it is supoosed to work like this. You need to add lines to the stepperdataX27 and the stepperX27 arrays. The entries need to be seperated by a comma.
I will add a comment to the skech to clarify this.
